Source/websocket/WebSocketLink.cpp

```diff
@@ -142,18 +136,22 @@ namespace Web {
                 if ((_controlStatus & REQUEST_CLOSE) != 0) {
                     dataFrame[result++] = FINISHING_FRAME | Protocol::CLOSE;
                     _controlStatus &= (~REQUEST_CLOSE);
-                    dataFrame[result++] = (_setFlags & MASKING_FRAME);
                 }
                 if (((_controlStatus & REQUEST_PING) != 0) && ((result + 1) < maxSendSize)) {
                     dataFrame[result++] = FINISHING_FRAME | Protocol::PING;
                     _controlStatus &= (~REQUEST_PING);
-                    dataFrame[result++] = (_setFlags & MASKING_FRAME);
                 }
                 if (((_controlStatus & REQUEST_PONG) != 0) && ((result + 1) < maxSendSize)) {
                     dataFrame[result++] = FINISHING_FRAME | Protocol::PONG;
                     _controlStatus &= (~REQUEST_PONG);
-                    dataFrame[result++] = (_setFlags & MASKING_FRAME);
                 }
+
+                dataFrame[result++] = (_setFlags & MASKING_FRAME);
+                // Now it seems only control message, hence append with masking keys
+                uint8_t maskKey[4];
```

Comment:
The following lines may only be executed when the MASKING_FRAME flag is set (so add an if(_setFlags & MASKING_FRAME) != 0) around it.

Also line 135 (in the new version) needs to be changed 
(this one:  if (((_controlStatus & (REQUEST_CLOSE | REQUEST_PING | REQUEST_PONG)) != 0) && ((result + 1) < maxSendSize))

the "result + 1" was always wrong I think and needs to be result + 2 but it needs to be result + 6 when  MASKING_FRAME flag is set to make sure there is enough room for it (so easiest I think is probably to use ((_setFlags & MASKING_FRAME) == 0) ? : ;

Reply:
done